core/templates/dev/head/domain/topic/TopicUpdateService.js

```diff
@@ -230,6 +228,61 @@ oppia.factory('TopicUpdateService', [
           // Get the current change list.
           var currentChangeList = UndoRedoService.getChangeList();
           var indicesToDelete = [];
+          var oldSubtopicIdSkillIdMap = {};
+          var index = null;
+          var skillIds = subtopic.getSkillIds();
+          for (var i = 0; i < currentChangeList.length; i++) {
+            var changeDict =
+              currentChangeList[i].getBackendChangeObject();
+            if (oldSubtopicIdSkillIdMap[changeDict.skill_id] === undefined) {
+              if (changeDict.new_subtopic_id === subtopicId) {
+                // Record the origin subtopic id for the skill.
+                oldSubtopicIdSkillIdMap[changeDict.skill_id] =
+                  changeDict.old_subtopic_id;
+                index = i;
+              }
+            } else {
+              if (changeDict.cmd === CMD_REMOVE_SKILL_ID_FROM_SUBTOPIC) {
+                // If the skill was removed from all subtopics at some point,
+                // modify that change dict, so that it was as if it was removed
+                // from its recorded origin.
+                changeDict.subtopic_id =
+                  oldSubtopicIdSkillIdMap[changeDict.skill_id];
+              } else {
+                // At every move operation, set old subtopic id as the recorded
+                // old subtopic id.
+                changeDict.old_subtopic_id =
+                  oldSubtopicIdSkillIdMap[changeDict.skill_id];
+              }
+              currentChangeList[i].setBackendChangeObject(changeDict);
+              // Remove the previous move operation that was related to the
+              // skill id since, the old subtopic id of the current change dict
+              // is set to the actual initial subtopic id.
+              currentChangeList.splice(index, 1);
+              i--;
+              index = i;
+            }
+          }
+          for (var i = 0; i < currentChangeList.length; i++) {
```

Comment:
Doing this in a separate for loop seems a bit icky. Shouldn't we just do it in the same for loop that handles subtopic deletion, and, at the point of subtopic deletion, move the skills back to where they last were and update the changelist accordingly?

Reply:
Done, removed this.